daemon/k8s_watcher.go

```diff
@@ -1449,6 +1449,30 @@ func getUpdatedCNPFromStore(ciliumV2Store cache.Store, cnp *cilium_v2.CiliumNetw
 	return serverRule, nil
 }
 
+func (d *Daemon) recalculatePolicyFromCNPUpdate(ciliumV2Store cache.Store, cnp *cilium_v2.CiliumNetworkPolicy) {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
+		logfields.K8sAPIVersion:           cnp.TypeMeta.APIVersion,
+		logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+	})
+
+	scopedLog.Infof("recalculating policy from CNP update")
+
+	rev := d.policy.GetRevision()
+	d.TriggerPolicyUpdates(false)
```

Comment:
What is the reason we need to trigger a policy update? The spec is guaranteed to be unchanged. I understand that the `rev` above is not the `rev` of the original CNP import but even then a, policy trigger does not seem required.

It would in fact be better to store the rev and status in a map indexed by `<namespace>/<name>` so we can store state per CNP and have a couple of fields such as `OK`, `Enforcing`, `Rev`. This field can be looked up from the controller and it can continuously sync it to k8s.

Reply:
@tgraf my main motivation here was maintaining uniformity between both updates of the same controller (`sync-cnp-policy-status`), as the issue which caused this ticket to be filed was due to us replacing the controller of the same name with different logic, regardless of if the prior controller failed or not. I agree that this is a 'kitchen-sink' approach - it seemed the safest thing to do remediate the current problem to unblock those who have been facing this. I will work on integrating your proposed changes.

Reply:
I've been thinking about this, and I think the easiest thing to do here would be to add functionality to basically 'trigger' the controller to run again - all of the required information we need right now is encapsulated within the controller itself. When we run it again, it will be able to get the most recent version of the CNP from the store (including the annotations), and we will never have to expose any other state to the daemon, as it would all be contained within the preexisting `DoFunc`.

Reply:
The nice thing about triggering the controller to run again is that the existing framework guarantees that we are not running multiple goroutines at the same time (no race conditions will occur) and we do not have to maintain any more shared state.

Reply:
After thinking about this a lot / trying some stuff out locally, I am against adding another cache of K8s resources, locally, because the Kubernetes store that we use already acts as a local cache. It guarantees that we do not push out-of-date resources to the cache, while maintaining a local cache does not maintain such an invariant. I've updated the code to simply grab the revision number from the `CiliumNetworkPolicyNodeStatus` for this node instead of updating the revision and triggering policy updates. This will maintain the invariant that the revision number that is that of the policy repository which first implemented the given CNP.